Java: model remaining subpackages of Apache Commons Collections

owen-mc opened the pull request:
Currently the tests are missing. I will add them after the models have been reviewed, to avoid having to update the tests for any models that are changed.

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".bidimap;AbstractDualBidiMap$BidiMapIterator;true;BidiMapIterator;;;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".bidimap;AbstractDualBidiMap$EntrySet;true;EntrySet;;;MapKey of Argument[0];MapKey of Element of Argument[-1];value",
+          ".bidimap;AbstractDualBidiMap$EntrySet;true;EntrySet;;;MapValue of Argument[0];MapValue of Element of Argument[-1];value",
+          ".bidimap;AbstractDualBidiMap$EntrySetIterator;true;EntrySetIterator;;;Element of Argument[0];MapKey of Element of Argument[-1];value",
```
```suggestion
          ".bidimap;AbstractDualBidiMap$EntrySetIterator;true;EntrySetIterator;;;MapKey of Element of Argument[0];MapKey of Element of Argument[-1];value",
```

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".bidimap;AbstractDualBidiMap$MapEntry;true;MapEntry;;;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".bidimap;AbstractDualBidiMap$MapEntry;true;MapEntry;;;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".bidimap;AbstractDualBidiMap$Values;true;Values;;;MapValue of Argument[0];Element of Argument[-1];value",
+          ".bidimap;AbstractDualBidiMap$ValuesIterator;true;ValuesIterator;;;Element of Argument[0];Element of Argument[-1];value",
```
We should be consistent about propagating or not from the `AbstractDualBidiMap<?,V> parent` parameter (EntrySetIterator currently does but ValuesIterator does not, for example). We probably should propagate (though it doesn't matter a lot since these are protected constructors and will be a pain to test also)

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".bidimap;DualLinkedHashBidiMap;true;DualLinkedHashBidiMap;;;MapValue of Argument[1];MapKey of Argument[-1];value",
+          ".bidimap;DualLinkedHashBidiMap;true;DualLinkedHashBidiMap;;;MapKey of Argument[2];MapValue of Argument[-1];value",
+          ".bidimap;DualLinkedHashBidiMap;true;DualLinkedHashBidiMap;;;MapValue of Argument[2];MapKey of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map,Map,BidiMap);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map,Map,BidiMap);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map,Map,BidiMap);;MapKey of Argument[1];MapValue of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map,Map,BidiMap);;MapValue of Argument[1];MapKey of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map,Map,BidiMap);;MapKey of Argument[2];MapValue of Argument[-1];value",
+          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;(Map,Map,BidiMap);;MapValue of Argument[2];MapKey of Argument[-1];value",
```
```suggestion
          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;;;MapKey of Argument[0];MapKey of Argument[-1];value",
          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;;;MapValue of Argument[0];MapValue of Argument[-1];value",
          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;;;MapKey of Argument[1];MapValue of Argument[-1];value",
          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;;;MapValue of Argument[1];MapKey of Argument[-1];value",
          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;;;MapKey of Argument[2];MapValue of Argument[-1];value",
          ".bidimap;DualTreeBidiMap;true;DualTreeBidiMap;;;MapValue of Argument[2];MapKey of Argument[-1];value",
```

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".bidimap;AbstractSortedBidiMapDecorator;true;AbstractSortedBidiMapDecorator;;;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".bidimap;DualHashBidiMap;true;DualHashBidiMap;;;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".bidimap;DualHashBidiMap;true;DualHashBidiMap;;;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".bidimap;DualHashBidiMap;true;DualHashBidiMap;;;MapKey of Argument[1];MapValue of Argument[-1];value",
```
These have `createBidiMap` methods too, though they're protected so don't matter much (but we are covering protected constructors, so might want to pick them up for consistency)

smowton reviewed (commented):
Note on protected constructors / static methods; these are probably of very marginal relevance for dataflow purposes, so I think it would be fine to either skip testing them and just make a best effort, or if need be drop the models.

atorralba commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".iterators;ObjectArrayListIterator;true;ObjectArrayListIterator;;;ArrayElement of Argument[0];Element of Argument[-1];value",
+          ".iterators;ObjectArrayListIterator;true;getArray;;;Element of Argument[-1];ArrayElement of ReturnValue;value",
+          ".iterators;ObjectGraphIterator;true;ObjectGraphIterator;(Iterator);;Element of Element of Argument[0];Element of Argument[-1];value",
+          ".iterators;PeekingIterator;true;peekingIterator;;;Element of Argument[0];Element of ReturnValue;value",
```
Missing the [constructor](https://commons.apache.org/proper/commons-collections/javadocs/api-4.4/org/apache/commons/collections4/iterators/PeekingIterator.html#PeekingIterator-java.util.Iterator-)?

atorralba commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".iterators;PeekingIterator;true;peekingIterator;;;Element of Argument[0];Element of ReturnValue;value",
+          ".iterators;PeekingIterator;true;peek;;;Element of Argument[-1];ReturnValue;value",
+          ".iterators;PeekingIterator;true;element;;;Element of Argument[-1];ReturnValue;value",
+          ".iterators;PermutationIterator;true;PermutationIterator;;;Element of Argument[0];Element of Element of Argument[-1];value",
```
This is smart :)

atorralba reviewed (commented):
Reviewed the `.collection` and `.iterators` packages.

About wrapper classes, I was wondering if we could model its "proxying" aspect and @joefarebrother mentioned that maybe that could be fixed using synthetic fields, which theoretically would enable backtracking side effects to the wrapped collection, e.g.

~~~java
List a = new ArrayList();
WrappedList b = wrap(a);
b.add(source());
sink(a.get(0));
~~~

But it adds a significant amount of effort, since all inherited methods from the wrapped class hierarchy would need to be remodeled, so I don't know if it's worth it.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;AbstractLinkedList;true;addNodeAfter;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNodeAfter;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;addNode;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNode;;;Element of Argument[0];Element of Argument[1];value",
```
Not sure how this affects `Argument[1]`. From the code, it does touch the linked nodes next/previous fields but that's no value flow

owen-mc replied:
See :fish:

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;AbstractLinkedList;true;addNode;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNode;;;Element of Argument[0];Element of Argument[1];value",
+          ".list;AbstractLinkedList;true;getNode;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".list;AbstractLinkedList;true;createSubListIterator;;;Element of Argument[-1];Element of ReturnValue;value",
```
This doesn't seem to carry anything from the qualifier, only from the subList (arg-0)

owen-mc replied:
I have deliberately done this so that we will still find the flow if we have not been able to track value flow into `arg[0]`. It does risk creating FPs.

owen-mc replied:
In the end I removed most of these models, because they are for private methods which I don't expect to be used as the public interface for any subclasses.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;AbstractLinkedList$Node;true;Node;(Node,Node,Object);;Element of Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList$Node;true;Node;(Node,Node,Object);;Argument[2];Element of Argument[-1];value",
+          ".list;AbstractLinkedList$Node;true;Node;(Node,Node,Object);;Argument[2];Element of Argument[0];value",
+          ".list;AbstractLinkedList$Node;true;Node;(Node,Node,Object);;Argument[2];Element of Argument[1];value",
```
Same as above, not sure how `value` affects prev and next nodes here

owen-mc replied:
See :fish:

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;SetUniqueList;true;asSet;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".list;SetUniqueList;true;createSetBasedOnList;;;Element of Argument[1];Element of ReturnValue;value",
+          ".list;TransformedList;true;TransformedList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;TransformedList;true;transformingList;;;Element of Argument[0];Element of ReturnValue;value",
```
missing `transformedList`

owen-mc replied:
The docs say "If there are any elements already in the list being decorated, they will be transformed by this method." So this can't be properly modelled until we have lambda flow.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;SetUniqueList;true;setUniqueList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;SetUniqueList;true;asSet;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".list;SetUniqueList;true;createSetBasedOnList;;;Element of Argument[1];Element of ReturnValue;value",
+          ".list;TransformedList;true;TransformedList;;;Element of Argument[0];Element of Argument[-1];value",
```
I think most of these should be `taint` and not `value` due to the transformer changing the items.

owen-mc replied:
I have only modelled methods which do not tranform the items. The other methods I have left until we have lambda flow.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          // Note that when lambdas are supported we should have more models for DefaultedMap, LazyMap, TransformedMap, TransformedSortedMap
+          ".map;AbstractHashedMap;true;AbstractHashedMap;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".map;AbstractHashedMap;true;AbstractHashedMap;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".map;AbstractHashedMap;true;convertKey;;;Argument[0];ReturnValue;value",
```
```suggestion
          ".map;AbstractHashedMap;true;convertKey;;;Argument[0];ReturnValue;taint",
```

owen-mc replied:
I was thinking about the implementation in this class, which does return `Argument[0]` except in the case that it is null. I shouldn't have put `true` for the `subtypes` column. Presumably most overrides would implement a hash algorithm, which would be very hard to reverse, so I'm not sure it's worth modelling this as taint flow for subtypes.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".map;AbstractHashedMap;true;entryKey;;;MapKey of Argument[0];ReturnValue;value",
+          ".map;AbstractHashedMap;true;entryValue;;;MapValue of Argument[-1];ReturnValue;value",
+          ".map;AbstractHashedMap;true;entryValue;;;MapValue of Argument[0];ReturnValue;value",
+          ".map;AbstractHashedMap;true;createEntrySetIterator;;;MapKey of Element of ReturnValue; MapKey of Argument[-1];value",
+          ".map;AbstractHashedMap;true;createEntrySetIterator;;;MapValue of Element of ReturnValue; MapValue of Argument[-1];value",
+          ".map;AbstractHashedMap;true;createKeySetIterator;;;Element of ReturnValue; MapKey of Argument[-1];value",
+          ".map;AbstractHashedMap;true;createValuesIterator;;;Element of ReturnValue; MapValue of Argument[-1];value",
```
Something looks odd here. Input has `ReturnValue`, output has spaces in it. And it seems like they're swapped

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".map;AbstractLinkedMap;true;AbstractLinkedMap;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".map;AbstractMapDecorator;true;AbstractMapDecorator;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".map;AbstractMapDecorator;true;AbstractMapDecorator;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".map;AbstractMapDecorator;true;decorated;;;MapKey of Argument[-1];MapKey of Argument[0];value",
+          ".map;AbstractMapDecorator;true;decorated;;;MapValue of Argument[-1];MapValue of Argument[0];value",
```
output should be `MapKey/Value of ReturnValue` no?

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".map;CompositeMap;true;CompositeMap;(Map,Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".map;CompositeMap;true;CompositeMap;(Map,Map);;MapKey of Argument[1];MapKey of Argument[-1];value",
+          ".map;CompositeMap;true;CompositeMap;(Map,Map);;MapValue of Argument[1];MapValue of Argument[-1];value",
+          ".map;CompositeMap;true;CompositeMap;(Map,Map,MapMutator);;MapKey of Argument[0];MapKey of Argument[-1];value",
```
I wonder if the variations including the `MapMutator` should be handled as `taint`

owen-mc replied:
My impression was that the `MapMutator` was intended to be used to decide which of the array of maps to do the mutation operation on, which we don't care about in our model.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".map;MultiValueMap;true;multiValueMap;;;Element of MapValue of Argument[0];Element of MapValue of ReturnValue;value",
+          ".map;MultiValueMap;true;getCollection;;;Element of MapValue of Argument[-1];Element of ReturnValue;value",
+          ".map;MultiValueMap;true;putAll;;;Argument[0];MapKey of Argument[-1];value",
+          ".map;MultiValueMap;true;putAll;;;Element of Argument[1];Element of MapValue of Argument[-1];value",
```
```suggestion
          ".map;MultiValueMap;true;putAll;;;Element of Argument[0];Element of MapValue of Argument[-1];value",
```

owen-mc replied:
I'm modelling the `putAll(K key, Collection<V> values)` method here, not `putAll(Map<? extends K,?> map)` (which is inherited from `Map`. I'll explicitly give the signature to make this clearer.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(ExpirationPolicy,Map);;MapValue of Argument[1];MapValue of Argument[-1];value",
+          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(long,Map);;MapKey of Argument[1];MapKey of Argument[-1];value",
+          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(long,Map);;MapValue of Argument[1];MapValue of Argument[-1];value",
+          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(long,TimeUnit,Map);;MapKey of Argument[1];MapKey of Argument[-1];value",
+          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(long,TimeUnit,Map);;MapValue of Argument[1];MapValue of Argument[-1];value",
```
```suggestion
          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(long,TimeUnit,Map);;MapKey of Argument[2];MapKey of Argument[-1];value",
          ".map;PassiveExpiringMap;true;PassiveExpiringMap;(long,TimeUnit,Map);;MapValue of Argument[2];MapValue of Argument[-1];value",
```

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".properties;AbstractPropertiesFactory;true;load;(Reader);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(String);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(URI);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(URL);;Argument[0];ReturnValue;taint"
```
Do we need to model `SortedProperties`. I couldn't find (yet) where we model `Properties` to account for that

owen-mc replied:
What would you model in `SortedProperties`? I don't think we model `Properties` for data flow at the moment. We probably should. We do model `Dictionary`, which is a superclass of it.

bmuskalla commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+/**
+ * Value-propagating models for the package `org.apache.commons.collections4.multiset`.
+ */
+private class ApacheMultiSetModel extends SummaryModelCsv {
```
I'm somehow missing `MultiSet` with methods like `uniqueSet`

owen-mc replied:
That is in the base package, rather than this subpackage. It is modelled in the class `ApacheCollectionsModel`.

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".set;CompositeSet$SetMutator;true;addAll;;;Element of Argument[2];Element of Argument[0];value",
+          ".set;CompositeSet$SetMutator;true;addAll;;;Element of Argument[2];Element of Element of Argument[1];value",
+          ".set;CompositeSet;true;CompositeSet;(Set);;Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;CompositeSet;(Set[]);;Element of Element of Argument[0];Element of Argument[-1];value",
```
ArrayElement

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".set;CompositeSet$SetMutator;true;addAll;;;Element of Argument[2];Element of Element of Argument[1];value",
+          ".set;CompositeSet;true;CompositeSet;(Set);;Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;CompositeSet;(Set[]);;Element of Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;addComposited;(Set);;Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;addComposited;(Set,Set);;Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;addComposited;(Set,Set);;Element of Argument[1];Element of Argument[-1];value",
```
Can discard the signatures, these are unambiguous

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".set;CompositeSet;true;addComposited;(Set);;Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;addComposited;(Set,Set);;Element of Argument[0];Element of Argument[-1];value",
+          ".set;CompositeSet;true;addComposited;(Set,Set);;Element of Argument[1];Element of Argument[-1];value",
+          ".set;CompositeSet;true;addComposited;(Set[]);;Element of Element of Argument[0];Element of Argument[-1];value",
```
ArrayElement

smowton commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".set;ListOrderedSet;true;asList;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".set;ListOrderedSet;true;get;;;Element of Argument[-1];ReturnValue;value",
+          ".set;ListOrderedSet;true;add;;;Argument[1];Element of Argument[-1];value",
+          ".set;ListOrderedSet;true;add;;;Element of Argument[1];Element of Argument[-1];value",
```
There's also an addAll op not specified on Collection

owen-mc replied:
It seems this line was meant to be `addAll`

owen-mc commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".collection;SynchronizedCollection;true;synchronizedCollection;;;Element of Argument[0];Element of ReturnValue;value",
+          ".collection;SynchronizedCollection;true;decorated;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".collection;TransformedCollection;true;TransformedCollection;;;Element of Argument[0];Element of Argument[-1];value",
+          ".collection;TransformedCollection;true;transformingCollection;;;Element of Argument[0];Element of ReturnValue;value",
```
The docs say "If there are any elements already in the collection being decorated, they will be transformed by this method." So this can't be properly modelled until we have lambda flow.

aschackmull replied:
We have lambda flow now.

owen-mc replied:
Good point. I think I would prefer to do the extra modelling enabled by lambda flow for this package as a separate task, after this PR has been merged.

owen-mc commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;AbstractLinkedList;true;removeLast;;;Element of Argument[-1];ReturnValue;value",
+          ".list;AbstractLinkedList;true;updateNode;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;updateNode;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;createNode;;;Argument[0];Element of ReturnValue;value",
```
I have chosen to model `AbstractLinkedList` in a slightly less precise way than you might expect. I have conflated the contents of any particular node with the contents of all the nodes. The problem this solves is that from any node you can get to any other node using `next` and `previous` (and I expect these methods to be commonly used).

owen-mc commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;AbstractLinkedList;true;updateNode;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;createNode;;;Argument[0];Element of ReturnValue;value",
+          ".list;AbstractLinkedList;true;addNodeBefore;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNodeBefore;;;Argument[1];Element of Argument[0];value",
```
See :fish:

owen-mc commented on java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll:
```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;AbstractLinkedList;true;addNodeBefore;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNodeBefore;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;addNodeAfter;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNodeAfter;;;Argument[1];Element of Argument[0];value",
```
See :fish:

owen-mc commented:
I've now added the tests (and stubs). I fixed some errors in models which became obvious on the way (including in old models for this package).

I had to force-push because I rebased on main to get some improvements to helper utilities, but I haven't changed the original four commits at all, so they shouldn't need to be re-reviewed.

atorralba commented on java/ql/test/library-tests/frameworks/apache-collections/test.ql:
```diff
@@ -1,16 +1,19 @@
         //"package;type;overrides;name;signature;ext;inputspec;outputspec;kind",
-        // This is temporarily modelled for the helper function newEnumerationWithElement, until the relevant package is modelled
-        "org.apache.commons.collections4.iterators;IteratorEnumeration;true;IteratorEnumeration;;;Element of Argument[0];Element of Argument[-1];value",
+        "generatedtest;Test;false;newWithElement;;;Argument[0];Element of ReturnValue;value",
+        "generatedtest;Test;false;getElement;;;Element of Argument[0];ReturnValue;value",
+        "generatedtest;Test;false;newWithArrayElement;;;Argument[0];ArrayElement of ReturnValue;value",
+        "generatedtest;Test;false;getArrayElement;;;ArrayElement of Argument[0];ReturnValue;value",
+        "generatedtest;Test;false;newWithMapValue;;;Argument[0];MapValue of ReturnValue;value",
+        "generatedtest;Test;false;getMapValue;;;MapValue of Argument[0];ReturnValue;value",
+        "generatedtest;Test;false;newWithMapKey;;;Argument[0];MapKey of ReturnValue;value",
+        "generatedtest;Test;false;getMapKey;;;MapKey of Argument[0];ReturnValue;value",
         "generatedtest;Test;false;newRBWithMapValue;;;Argument[0];MapValue of ReturnValue;value",
         "generatedtest;Test;false;newRBWithMapKey;;;Argument[0];MapKey of ReturnValue;value"
```
Are these models necessary? The methods seem to have actual implementations that should be already covered by the dataflow library.

smowton replied:
I checked and only two are necessary, the ones constructing ResourceBundle (looks like we don't model that yet). I've pushed a commit removing those not required.
